crates/legacy/src/service.rs

```diff
@@ -1008,25 +675,63 @@ where
         num_nodes: usize,
     ) -> Result<AvailableBlockData<Types>, BuildError> {
         // Update the stored `num_nodes` with the given value, which will be used for VID computation.
-        self.global_state.write_arc().await.num_nodes = num_nodes;
+        trace!(
+            new_num_nodes = num_nodes,
+            old_num_nodes = self.num_nodes.load(Ordering::Relaxed),
+            "Updating num_nodes"
+        );
+
+        self.num_nodes.store(num_nodes, Ordering::Relaxed);
 
         self.claim_block(block_hash, view_number, sender, signature)
             .await
     }
 
+    #[tracing::instrument(skip_all)]
     async fn claim_block_header_input(
         &self,
         block_hash: &BuilderCommitment,
         view_number: u64,
         sender: Types::SignatureKey,
         signature: &<<Types as NodeType>::SignatureKey as SignatureKey>::PureAssembledSignatureType,
     ) -> Result<AvailableBlockHeaderInput<Types>, BuildError> {
-        Ok(self
-            .claim_block_header_input_implementation(block_hash, view_number, sender, signature)
-            .await?)
-    }
+        let start = Instant::now();
+        // verify the signature
+        if !sender.validate(signature, block_hash.as_ref()) {
+            warn!("Signature validation failed in claim_block_header_input");
+            return Err(Error::<Types>::SignatureValidation.into());
+        }
+
+        let block_id = BlockId {
+            hash: block_hash.clone(),
+            view: Types::View::new(view_number),
+        };
+
+        trace!("Processing claim_block_header_input request");
+
+        let (truncated, info) = timeout(
+            self.max_api_waiting_time,
+            self.claim_block_header_input_implementation(block_id),
+        )
+        .await
+        .inspect_err(|_| {
+            // we can't keep up with this block size, reduce max block size
+            self.block_size_limits.decrement_block_size();
+        })
+        .map_err(|_| Error::<Types>::ApiTimeout)??;
+
+        if self.max_api_waiting_time.saturating_sub(start.elapsed())
+            > self.max_api_waiting_time / VID_RESPONSE_TARGET_MARGIN_DIVISOR
```

Comment:
Nit: Why do we use a VID-related const? Or should we rename it?

Reply:
Another nit: We use different divisors (2 or 10) at other places. Can we add some consts for them as well?

Reply:
Can't think of any name particularly better, so let's keep it as is!